TST use global_random_seed in sklearn/metrics/tests/test_regression.py

sortofamudkip opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards #22827. 

#### What does this implement/fix? Explain your changes.

The modified tests are:
* `test_tweedie_deviance_continuity`: **tolerance was changed from 1e-6 to 1e-5** to allow all tests to pass.
* `test_mean_absolute_percentage_error`: all tests passed.
* `test_dummy_quantile_parameter_tuning`: all tests passed.
* `test_pinball_loss_relation_with_mae`: all tests passed.
* `test_mean_pinball_loss_on_constant_predictions`: **some tests failed.**

In particular, tests failed locally with the following configurations:

|                            | `n_samples=3000` | `n_samples=30000` |
|----------------------------|------------------|-------------------|
| `distribution=uniform`     | 6/300 failed     | 0/300 failed      |
| `distribution=exponential` | 6/300 failed     | 0/300 failed      |
| `distribution=normal`      | 53/300 failed    | 117/300 failed    |

All tests for `distribution=lognormal` passed.

Increasing the number of samples causes `distribution=normal` to fail *more* often.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

sortofamudkip committed:
add global_random_seed to test_regression.py [all random seeds]
test_tweedie_deviance_continuity
test_mean_absolute_percentage_error
test_mean_pinball_loss_on_constant_predictions
test_dummy_quantile_parameter_tuning
test_pinball_loss_relation_with_mae

sortofamudkip committed:
Merge branch 'main' into rng_test_regression

sortofamudkip committed:
Merge branch 'scikit-learn:main' into rng_test_regression

OmarManzoor commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -583,19 +585,20 @@ def objective_func(x):
     assert result.success
     # The minimum is not unique with limited data, hence the large tolerance.
-    assert result.x == pytest.approx(best_pred, rel=1e-2)
+    assert result_x == pytest.approx(best_pred, rel=1e-2)
```
I think the issue is mostly with recovering the actual value which can turn out to be somewhat different than the actual best_pred. Moreover the optimization might also not be converging to the actual minimum loss on some occasions using the default tolerance in the minimize method. 
What would be the best approach to handling this @ogrisel @lorentzenchr

jeremiedbb replied:
Given the approximate nature of this part of the test due to the limited data, increasing the tol looks fine to me. I also added an absolute tol as well to take into account the normal distribution + 0.5 quantile (i.e. expected close to 0).

OmarManzoor reviewed (commented):
Thanks for the PR @sortofamudkip

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/sortofamudkip/30865

jeremiedbb committed:
increase tol

jeremiedbb committed:
comment

jeremiedbb committed:
[all random seeds]
test_tweedie_deviance_continuity
test_mean_absolute_percentage_error
test_mean_pinball_loss_on_constant_predictions
test_dummy_quantile_parameter_tuning
test_pinball_loss_relation_with_mae

jeremiedbb reviewed (approved):
LGTM. Thanks @sortofamudkip.
